### PR TITLE
Support partition schema evolution for ORC and Parquet

### PR DESCRIPTION
**Pull request by kokosing**

### CODE REVIEW

File: presto-hive/src/main/java/io/prestosql/plugin/hive/HiveSplitManager.java [hunk 1]
```diff
@@ -360,30 +365,71 @@ public CounterStat getHighMemorySplitSource()
         return concat(partitionBatches);
     }
 
-    private TableToPartitionMapping getTableToPartitionMapping(SchemaTableName tableName, String partName, List<Column> tableColumns, List<Column> partitionColumns)
+    private TableToPartitionMapping getTableToPartitionMapping(ConnectorSession session, SchemaTableName tableName, String partName, List<Column> tableColumns, List<Column> partitionColumns)
```
[the remark this answers is not in the document]
Comment (kokosing) [reply presto-hive/src/main/java/io/prestosql/plugin/hive/HiveSplitManager.java]: I would leave it as it is. It is quite simple code.

File: presto-product-tests/src/main/java/io/prestosql/tests/hive/TestHivePartitionSchemaEvolution.java [hunk 2]
```diff
@@ -0,0 +1,149 @@
...
+            // partition is dropping a column at the start
+            testEvolution(table, "ALTER TABLE %s REPLACE COLUMNS (tiny_column tinyint, int_column int, float_column float, varchar_column varchar(20))", row(null, 1, 1.1, "jeden", 1));
+
+            // partition is dropping a column in the middle
+            testEvolution(table, "ALTER TABLE %s REPLACE COLUMNS (int_column int, tiny_column tinyint, float_column float, varchar_column varchar(20))", row(1, null, 1.1, "jeden", 1));
+
+            // partition is dropping a column at the end
+            testEvolution(table, "ALTER TABLE %s REPLACE COLUMNS (int_column int, float_column float, varchar_column varchar(20), tiny_column tinyint)", row(1, 1.1, "jeden", null, 1));
+
+            // partition is dropping a column at the start
```
Comment (phd3) [inline presto-product-tests/src/main/java/io/prestosql/tests/hive/TestHivePartitionSchemaEvolution.java]: redundant comment

File: presto-product-tests/src/main/java/io/prestosql/tests/hive/TestHivePartitionSchemaEvolution.java [hunk 3]
```diff
@@ -0,0 +1,149 @@
...
+                log.warn("Unable to execute: %s, due: %s", sql, message);
+                return false;
+            }
+            throw e;
+        }
+    }
+
+    private static TemporaryHiveTable createTable(String format)
+    {
+        String tableName = "parquet_schema_evolution_" + randomTableSuffix();
```
Comment (phd3) [inline presto-product-tests/src/main/java/io/prestosql/tests/hive/TestHivePartitionSchemaEvolution.java]: use format instead of parquet for tableName

**Comment by kokosing** [comment]
@phd3 Comments addressed. Thank you for the review!

**Comment by phd3** [comment]
Looks good! Thanks for implementing this change @kokosing

**Comment by rclaude** [comment]
Hi @kokosing 
thanks a lot for this work!

Is the overall design of this change documented somewhere to understand the key points and maybe limitations (fields can be added,removed,re-ordered ?)

I implemented something similar (but for parquet only) and had to rely on (and patch) ParquetReader:
1st part of the patch (Parquet Reader): https://github.com/criteo-forks/presto/commit/bf80c47621a03d13367579990bf6dc9c571786fc
2nd part (schema evolution):  https://github.com/criteo-forks/presto/commit/82e55fc628e4e067dd44b0b9b06dc63b91df2254

It seem your design is simpler / better + has unit tests and I'm trying to understand it before throwing away the patches I have in my fork :) !

**Comment by phd3** [comment]
@rclaude I believe the patch behaves as follows when `BY_NAME` strategy is used: (@kokosing feel free to correct)

- For top level columns: they can be reordered, added/dropped in the middle (between partition and table schemas)

- Since the name based partition schema evolution should always be used with `use-column-names` property of the ORC/Parquet formats, the evolutions mentioned above are also supported between partition and file

- For struct fields, the hive coercion policy only allows fields being dropped/added at the end. So even though ORC/Parquet Readers access subfields in a name-based fashion, reordering or adding/dropping subfields in the middle is not supported overall.

**Comment by rclaude** [comment]
thanks a lot @phd3 for this summary!

In my company ETLs, one table often become nested in a downstream table. So allowing add/drop in the middle & re-order of top level columns require supporting those also for nested fields.

I'll have to keep my patches to support schema evolution in nested fields struct/array/map.

Maybe I can contribute that after rebasing on the changes made here ?

A quick summary of the approach i used :

There is 4 schemas: 
* file schema
* partition schema
* table schema
* schema used by the parquet reader

The key idea is to give the parquet reader a schema built by: 
Using the table schema and replacing the primitive types by the ones of the partition schema. 
Thus every partition data will be read with the table structure (with NULLs filled in for missing nested field/struct/array/map)

The schema given to the parquet reader doesn't necessarily match the file schema (this has never been a guaranty anyway) so the parquet reader has to cope with missing fields in the file:
https://github.com/criteo-forks/presto/commit/bf80c47621a03d13367579990bf6dc9c571786fc

This methodology can be followed for ORC too. 
I tried other approaches but found this one to be the more minimalist / maintainable that I could think of.

What do you think / are you interested in having this in upstream Presto ?

**Comment by findepi** [comment]
@rclaude yes, we definitely should pursue schema evolution for nested fields too.
As I see, you tried to cover this on the file level. We should also do this on partition level.
And, unfortunately, we need to maintain compaitbility with other tools in the ecosystem, so we need to understand eg
how hive 2 and 3 deals with this. Maybe we could compare with Athena too.

Did you try to research the compatibility?
Would you mind trying adding a test? (in `AbstractTestParquetReader` for file-level and `TestHiveIntegrationSmokeTest` otherwise)

**Comment by rclaude** [comment]
> As I see, you tried to cover this on the file level. We should also do this on partition level.
I'm not sure I understand what you mean but realized I forgot to mention an important detail:
It works at the partition level in the HiveSplitManager.  https://github.com/criteo-forks/presto/commit/82e55fc628e4e067dd44b0b9b06dc63b91df2254 (naming isn't great: middleground refer to the mixed schema explained in my previous comment)
It had to be partly implemented in the ParquetReader because at type coercion time the information about how many rows exist cannot be made up to fill in the correct numbers of NULLs.

= Compatibility =

== Hive ==
We implemented the same behavior in our [Hive fork](https://github.com/criteo-forks/hive/) (Finally! It was a todo left there by my team 7 years ago).
Unfortunately, we run a very custom Hive1 version with lots of back-ports from Hive2 & 3 so haven't pushed it upstream.
It might happen in the future: we will port our schema evolution patch on a Hive2 branch in the coming months.

== Athena ==
I don't know

= Test =
Those are clearly missing in the two commits I linked. 
I started working on the AbstractTestParquetReader ones but since it doesn't follow the same "read what you write" pattern as the other tests it was time consuming and I haven't finished that yet .. I'll do so before pushing a pull request.
For the real schema evolution tests: I have a suite of test scenarios that were done in an adhoc code base. I haven't learned how to work with Presto's docker based testing suite yet, but thanks to kokosing work, most of the heavy-lifting code is already there now!

**Pull request closed without merging**